design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,174 @@
+individual items from the queue at any position and not merely grab the top/bottom item from the 
+queue.
+
+The minimal information each item in the queue needs is only the OSD ID. The OSD Deployment managed
+by Rook can easily be inferred from the OSD ID.
+
+It will also be useful to keep a record of the number of times each OSD has be queued to determine 
+when Rook should stop attempting to update the OSD deployment. This can be attached to each queue
+item in the update queue via a struct.
+
+
+### User configuration
+
+#### Modifications to `CephCluster` CRD
+Establish a new `updatePolicy` section in the `CephCluster` `spec`. In this section, users can
+set options for how OSDs should be updated in parallel. Additionally, we can move some existing
+one-off configs related to updates to this section for better coherence. This also allows for
+a natural location where future update options can be added.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+# ...
+spec:
+  # ...
+  # Move these to the new updatePolicy but keep them here for backwards compatibility.
+  # These can be marked deprecated, but do not remove them until CephCluster CRD v2.
+  skipUpgradeChecks:
+  continueUpgradeAfterChecksEvenIfNotHealthy:
+  removeOSDsIfOutAndSafeToDestroy:
+  
+  # Specify policies related to updating the Ceph cluster and its components. This applies to
+  # minor updates as well as upgrades.
+  updatePolicy:
+    # skipUpgradeChecks is merely relocated from spec
+    skipUpgradeChecks: <bool, default=false>
+
+    # continueUpgradeAfterChecksEvenIfNotHealthy is merely relocated from spec
+    continueUpgradeAfterChecksEvenIfNotHealthy: <bool, default=false, relocated from spec>
+    
+    # allow for future additions to updatePolicy like healthErrorsToIgnore
+    
+    # Update policy for OSDs.
+    osds:
+      # removeIfOutAndSafeToDestroy is merely relocated from spec (removeOSDsIfOutAndSafeToRemove)
+      removeIfOutAndSafeToDestroy: <bool, default=false>
+
+      # Max number of OSDs to update at once. Rook will try to update this many OSDs at once if
+      # it is safe to do so, or fewer OSDs if it would be unsafe to update maxInParallel at once.
+      # This can be an absolute number or a percentage of total OSDs in the Ceph cluster.
+      # Rook defaults to updating 15% of OSDs in the cluster simultaneously if this value is unset.
+      # Inspired by Kubernetes apps/v1 RollingUpdateDeployment.MaxUnavailable.
+      # Note: I think we can hide the information about CRUSH from the user since it is not
+      #       necessary for them to understand that complexity.
+      maxInParallel: <k8s.io/apimachinery/pkg/util/intstr.intOrString, default=14%>
+      
+      # If updating OSDs takes longer than this amount of time, Rook will temporarily pause OSD
+      # updates to ensure other changes to the CephCluster resource are handled.
+      updateTimeout: <string, default=15m>
```

Comment:
I don't think we need a setting for this with the checkpoint strategy separately mentioned.

Reply:
We do still probably want a timeout to allow reconciliation to move on to updating "child" resources though, right?

Reply:
I thought we allowed the child controllers to update even while the OSDs were still updating, but maybe that's only triggered in parallel if the operator is restarted, instead of just updating the CephCluster CR with the upgraded image. Seems like we should notify the child controllers after the mons are updated, but in parallel of the OSDs. @leseb thoughts?

Reply:
I'll double check the code.

Reply:
The arbiter isn't configured until after OSD orchestration. But CephFS, etc. have their own reconciles that are separate.

Reply:
The child CRDs are triggered once the **monitors** are done updating, so effectively **before** the OSDs. Thus this flag is probably not necessary and could internally be managed.

Reply:
Should we do the same with the arbiter @travisn ?

Reply:
Hmm... okay. Does that definitely need to happen after OSDs are created.